### PR TITLE
This removes the crazy alignment requirements.

### PR DESCRIPTION
**Pull request by lemire**
This is simpler than https://github.com/simdjson/simdjson/pull/1068

Would help fix https://github.com/simdjson/simdjson/issues/1066

### CODE REVIEW

**Comment by lemire** [comment]
@jkeiser In theory, the misalignment of the SIMD loads (with AVX) can/should slightly lower the performance. But I have been writing for years that it is a secondary concern...

https://lemire.me/blog/2012/05/31/data-alignment-for-speed-myth-or-reality/

**Comment by lemire** [comment]
The performance is unchanged, except that I see some crazily slows allocation speeds for small files...

```
$ ./run './buildcandidate/benchmark/parse candidate/simdjson/jsonexamples/tree-pretty.json'
number of iterations 200

candidate/simdjson/jsonexamples/tree-pretty.json
================================================
      539 blocks -      34556 bytes -  3958 structurals ( 11.5 %)
special blocks with: utf8         0 (  0.0 %) - escape         0 (  0.0 %) - 0 structurals         0 (  0.0 %) - 1+ structurals       540 (100.0 %) - 8+ structurals       223 ( 41.3 %) - 16+ structurals         2 (  0.4 %)
special block flips: utf8         0 (  0.0 %) - escape         0 (  0.0 %) - 0 structurals         0 (  0.0 %) - 1+ structurals         0 (  0.0 %) - 8+ structurals       226 ( 41.9 %) - 16+ structurals         4 (  0.7 %)

All Stages
|    Speed        :  36.5315 ns per block ( 95.06%) -   0.5709 ns per byte -   4.9841 ns per structural -   1.7517 GB/s
|    Cycles       : 107.8889 per block    ( 98.83%) -   1.6860 per byte    -  14.7196 per structural    -    2.953 GHz est. frequency
|    Instructions : 363.6667 per block    ( 99.98%) -   5.6829 per byte    -  49.6160 per structural    -    3.371 per cycle
|    Misses       :      81 branch misses ( 96.11%) - 0 cache misses (  -nan%) - 0.00 cache references
|- Allocation
|    Speed        :   6.7352 ns per block ( 17.53%) -   0.1052 ns per byte -   0.9189 ns per structural -   9.5012 GB/s
|    Cycles       :   6.1167 per block    (  5.60%) -   0.0956 per byte    -   0.8345 per structural    -    0.908 GHz est. frequency
|    Instructions :   3.4389 per block    (  0.95%) -   0.0537 per byte    -   0.4692 per structural    -    0.562 per cycle
|    Misses       :      10 branch misses ( 11.87%) - 0 cache misses (  -nan%) - 0.00 cache references
|- Stage 1
|    Speed        :  13.0056 ns per block ( 33.84%) -   0.2032 ns per byte -   1.7744 ns per structural -   4.9204 GB/s
|    Cycles       :  44.6704 per block    ( 40.92%) -   0.6981 per byte    -   6.0945 per structural    -    3.435 GHz est. frequency
|    Instructions : 141.6111 per block    ( 38.93%) -   2.2129 per byte    -  19.3204 per structural    -    3.170 per cycle
|    Misses       :       8 branch misses (  9.49%) - 0 cache misses (  -nan%) - 0.00 cache references
|- Stage 2
|    Speed        :  16.4204 ns per block ( 42.73%) -   0.2566 ns per byte -   2.2403 ns per structural -   3.8971 GB/s
|    Cycles       :  56.1241 per block    ( 51.41%) -   0.8770 per byte    -   7.6572 per structural    -    3.418 GHz est. frequency
|    Instructions : 218.6167 per block    ( 60.10%) -   3.4163 per byte    -  29.8264 per structural    -    3.895 per cycle
|    Misses       :      64 branch misses ( 75.94%) - 0 cache misses (  -nan%) - 0.00 cache references

Warning: The processor frequency fluctuates in an expected way!!!
Expect the overall speed not to match stage 1 and stage 2 speeds.
Range for stage 1 and stage 2 : [3.418 GHz, 3.435 GHz], overall: 2.953 GHz.

50691.9 documents parsed per second (best)
```

**Comment by lemire** [comment]
Counterintuitively this PR seems to increase drastically the allocation cost.

File: include/simdjson/inline/parser.h [hunk 1]
```diff
@@ -101,19 +93,14 @@ inline simdjson_result<document_stream> parser::load_many(const std::string &pat
...
-    buf = (uint8_t *)internal::allocate_padded_buffer(len);
-    if (buf == nullptr)
-      return MEMALLOC;
-    memcpy((void *)buf, tmp_buf, len);
-  }
-
-  _error = implementation->parse(buf, len, doc);
-  if (realloc_if_needed) {
-    aligned_free((void *)buf); // must free before we exit
+    tmp_buf.reset((uint8_t *)internal::allocate_padded_buffer(len));
```
Comment (pps83) [inline include/simdjson/inline/parser.h:99]: With the change `allocate_padded_buffer` requires delete[] to be used to free the memory, eg `std::unique_ptr<uint8_t[]>` is ok here. The problem might happen if  `allocate_padded_buffer` is updated later to use malloc for example.

**Comment by pps83** [comment]
> Counterintuitively this PR seems to increase drastically the allocation cost.

Could it be that the cost is due to `delete[]` vs `free`?

**Comment by lemire** [comment]
@pps83 

> Could it be that the cost is due to delete[] vs free?

I don't know. I am really puzzled. The change in this PR should have, if anything, lowered, the allocation cost. There is clearly something I do not understand.

**Comment by pps83** [comment]
> @pps83
> 
> > Could it be that the cost is due to delete[] vs free?
> 
> I don't know. I am really puzzled. The change in this PR should have, if anything, lowered, the allocation cost. There is clearly something I do not understand.

delete[] internally has to track count of objects to properly delete each one of them. For plain types it's not really needed, but that's possibly the reason why it might be slow if something tries to malloc/free lots of data

File: include/simdjson/dom/parser.h [hunk 2]
```diff
@@ -395,7 +395,7 @@ class parser {
   // older versions of Visual Studio lack proper support for unique_ptr.
   std::unique_ptr<char[]> loaded_bytes;
   #else
-  std::unique_ptr<char[], decltype(&aligned_free_char)> loaded_bytes;
+  std::unique_ptr<char[]> loaded_bytes;
```
[the remark this answers is not in the document]
Comment (lemire) [reply include/simdjson/dom/parser.h:398]: Thanks!

File: include/simdjson/inline/parser.h (hunk 1 above)
Comment (lemire) [reply include/simdjson/inline/parser.h:99]: Of course.

**Comment by lemire** [comment]
> Could it be that the cost is due to delete[] vs free? Try to update the code to use malloc/free instead to see if that's the reason for the extra cost

I don't think so. There should be very little difference between the two. They do the same work.

I think the issue has to do with the memory allocation heuristics that get triggered by posix_memalign versus "new/malloc". I think I resolved the issue by always allocating at least 64KB (in the benchmark, I am not changing the core library). Anyhow, it is not very realistic to allocate memory for the parser in tiny amounts, you probably want 64KB in practice at a minimum.

This is the part that we do not control. Memory allocation happens at the runtime library level. There are system specific knobs that we could turn, but they are system specific.

**Comment by lemire** [comment]
> delete[] internally has to track count of objects to properly delete each one of them. For plain types it's not really needed, but that's possibly the reason why it might be slow if something tries to malloc/free lots of data

We do not benchmark delete, and if it were significantly slower than free, it would be a bug that we should report to the runtime libraries. It is effectively equivalent.

**Comment by lemire** [comment]
This should not drastically affect the throughput. I do not understand the failing tests.

File: include/simdjson/inline/padded_string.h [hunk 3]
```diff
@@ -21,17 +21,10 @@ inline char *allocate_padded_buffer(size_t length) noexcept {
   // return (char *) malloc(length + SIMDJSON_PADDING);
```
[the remark this answers is not in the document]
Comment (lemire) [reply include/simdjson/inline/padded_string.h:23]: You are quite right!

**Comment by lemire** [comment]
Overriding the perf checks. This is fine and a nice simplification.

**Pull request merged**